servers.go

```diff
@@ -189,105 +190,127 @@ type ComputedDistance struct {
 		}
+	}
 
-		var asn db.ASN
-
-		if r.asnDB != nil {
-			err = r.asnDB.Lookup(ip, &asn)
+	ruleInput := RuleInput{
+		IP:       ip.String(),
+		ASN:      asn,
+		Location: city,
+	}
 
-			if err != nil {
-				log.WithError(err).Warning("Unable to load ASN information")
-				return nil, -1, err
-			}
+	validServers := lo.Filter(s, func(server *Server, _ int) bool {
+		if !server.Available || !lo.Contains(server.Protocols, scheme) {
+			return false
 		}
-
-		ruleInput := RuleInput{
-			IP:       ip.String(),
-			ASN:      asn,
-			Location: city,
+		if len(server.Rules) > 0 && !server.checkRules(ruleInput) {
+			log.WithField("host", server.Host).Debug("Skipping server due to rules")
+			return false
 		}
+		return true
+	})
+	
+	if len(validServers) < 2 {
+		validServers = s
+	}
+	localServers := lo.Filter(validServers, func(server *Server, _ int) bool {
+		return server.Country == clientCountry
+	})
 
-		// First, filter our servers to what are actually available/match.
-		validServers := lo.Filter(s, func(server *Server, _ int) bool {
-			if !server.Available || !lo.Contains(server.Protocols, scheme) {
-				return false
-			}
-
-			if !server.checkRules(ruleInput) {
-				log.WithField("host", server.Host).Debug("Skipping server due to rules")
-				return false
-			}
-
-			return true
-		})
-
-		// Then, map them to distances from the client
-		c := lo.Map(validServers, func(server *Server, _ int) ComputedDistance {
-			distance := Distance(city.Location.Latitude, city.Location.Longitude, server.Latitude, server.Longitude)
+	const sameCityThreshold = 20000000.0 // 200 km
 
+	if len(localServers) > 0 {
+		computedLocal := lo.Map(localServers, func(server *Server, _ int) ComputedDistance {
+			d := Distance(city.Location.Latitude, city.Location.Longitude, server.Latitude, server.Longitude)
 			return ComputedDistance{
 				Server:   server,
-				Distance: distance,
+				Distance: d,
 			}
 		})
-
-		// Sort by distance
-		sort.Slice(c, func(i int, j int) bool {
-			return c[i].Distance < c[j].Distance
+		sort.Slice(computedLocal, func(i, j int) bool {
+			return computedLocal[i].Distance < computedLocal[j].Distance
 		})
-
+		if computedLocal[0].Distance < sameCityThreshold {
+			chosen := computedLocal[0]
+			r.serverCache.Add(cacheKey, chosen)
+			return chosen.Server, chosen.Distance, nil
+		}
 		choiceCount := r.config.TopChoices
-
-		if len(c) < r.config.TopChoices {
-			choiceCount = len(c)
+		if len(computedLocal) < choiceCount {
+			choiceCount = len(computedLocal)
 		}
-
-		log.WithFields(log.Fields{"count": len(c)}).Debug("Picking from top choices")
-
 		choices := make([]randutil.Choice, choiceCount)
-
-		for i, item := range c[0:choiceCount] {
+		for i, item := range computedLocal[0:choiceCount] {
 			choices[i] = randutil.Choice{
 				Weight: item.Server.Weight,
 				Item:   item,
 			}
 		}
-
-		choiceInterface = choices
-
-		r.serverCache.Add(scheme+"_"+ip.String(), choiceInterface)
+		r.serverCache.Add(cacheKey, choices)
```

Comment:
I ain't sure about here. I think we should just put the selected one to the cache. Will []randutil.Choice be asserted to ComputedDistance?